test/server.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package test supplies a fake Mixer server for use in testing. It should NOT
+// be used outside of testing contexts.
+package test // import "istio.io/mixer/test"
+
+import (
+	"fmt"
+	"net"
+
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	mixerpb "istio.io/api/mixer/v1"
+	"istio.io/api/mixer/v1/attributes"
+	"istio.io/mixer/pkg/attribute"
+)
+
+// AttributesServer implements the Mixer API to send mutable attributes bags to
+// a channel upon API requests. This can be used for tests that want to exercise
+// the Mixer API and validate server handling of supplied attributes.
+type AttributesServer struct {
+	GlobalDict map[string]int32
+	Attributes chan attribute.Bag
```

Comment:
we need to know which calls the attribute is coming from; check, report or quota.   We may need 3 channels, one for each.
Also for each call,  testing code may want to set response status too.  some calls are failed with specific error status.

For testing client cache,  testing code also need to return RelevantAttributes to tell client which attributes to use for cache.